Cuda persistent softmax

thorjohnsen opened the pull request:
Adds persistent cuda kernels that speed up SoftMax applied over the fast dimension, i.e. torch.nn.Softmax(dim=-1) and torch.nn.LogSoftmax(dim=-1). When the size is <= 1024, this code is 2-10x faster than the current code, speedup is higher for smaller sizes. This code works for half, float and double tensors with 1024 or fewer elements in the fast dimension. Numerical accuracy is on par with the current code, i.e. relative error is ~1e-8 for float tensors and ~1e-17 for double tensors. Relative error was computed against the CPU code.

The attached image shows kernel time in us for torch.nn.Softmax(dim=-1) applied to a half precision tensor of shape [16384,n], n is plotted along the horizontal axis. Similar uplifts can be seen for the backward pass and for LogSoftmax.

![image](https://user-images.githubusercontent.com/41591019/58212822-b63ebb00-7cb5-11e9-910d-1fc7d8585d58.png)

thorjohnsen committed:
Draft release of persistent softmax for CUDA

thorjohnsen committed:
Delete un-necessary test file

thorjohnsen committed:
Fix accuracy issue for double tensors

thorjohnsen committed:
Convert tabs to spaces

thorjohnsen committed:
Change __shfl_xor_sync to WARP_SHFL_XOR

soumith commented:
this is really cool!

ngimel commented:
Test failures are real.

ngimel commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,385 @@
+
+// WARP_BATCH number of batches.
+// WARP_ITERATOINS The number of iterations required for one warp to iterate over all data.
+// WARP_SIZE number of elements working on a single batch, has to be a power of two.
```
number of threads working on a single sample?

ngimel commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,385 @@
+    for (int offset = WARP_SIZE / 2; offset > 0; offset /= 2) {
+        acc_t val[WARP_BATCH];
+        #pragma unroll
+        for (int i = 0;i < WARP_BATCH;++i) {
```
Do 2 loops here provide perf benefits, or can they be fused like?
```
val = WARP_SHFL_XOR(..)
max_value[i] = <ternary operator>
```
If 2 loops are indeed necessary, comment might be in order.

ngimel commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,385 @@
+
+    // reduction sum
+    #pragma unroll
+    for (int offset = WARP_SIZE / 2; offset > 0; offset /= 2) {
```
I think it would make sense to have a __ device __ function for warp reduce, as it is used a few times. It could also handle max reduction.

thorjohnsen commented:
Yeah. They are intermittent. I ran one of the failed tests (test_softmax_dtype) two times in a row. It failed the first time and passed the second time. I'm looking into it.

ngimel commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,385 @@
+
+    switch (log2_elements) {
+    case 0: // 1
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,1, is_log_softmax>;
```
It looks to me like all the template arguments can be constexprs in the kernel, and the only real template argument is `next_power_of_2`. E.g.
```
WARP_SIZE = (np2 <  = 32) ? np2 : 32;
WARP_ITERATIONS = np2/WARP_SIZE;
WARP_BATCH = yet another ternary operator
```
This will make dispatch for forward and backward cleaner.

ngimel commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,385 @@
+}
+
+template<typename input_t, typename output_t, typename acc_t, bool is_log_softmax>
+bool dispatch_softmax(output_t *dst, const input_t *src, int softmax_elements, int softmax_elements_stride, int batch_count)
```
return value of dispatch_softmax is never checked, so you might as well make it void.

thorjohnsen replied:
I think I prefer to assert on the return value instead. dispatch_softmax will return false if you call it for a case that it can't handle properly.

thorjohnsen committed:
Fix embarrasing initialization bug

ngimel commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,385 @@
+        kernel<<<blocks, threads>>>(dst, src, batch_count, softmax_elements_stride, softmax_elements);
+        return true;
+    }
+    return false;
```
might make sense to do TORCH_INTERNAL_ASSERT on softmax_elements <=1024.

thorjohnsen replied:
That's actually a better idea than asserting on the return value from dispatch_softmax. Will do.

thorjohnsen commented:
I'm pretty sure I introduced this bug when I did the code cleanup, the input arrays were no longer being properly initialized. I fixed it, hopefully the tests will pass now.

ngimel commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,385 @@
+    } else if (softmax_elements <= 1024) {
+        // compute function index. there's a function for each power of two size up to 1024.
+        int log2_elements = 0;
+        while ((1 << log2_elements) < softmax_elements) ++log2_elements;
```
@ezyang, @mcarilli the functions for next power of 2 are often needed - does it make sense to make those helper functions generally available?

ngimel commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,390 @@
+    }
+
+    // reduction sum
+    constexpr uint32_t FULL_MASK = 0xffffffff;
```
same comment about FULL_MASK

ngimel commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,390 @@
+        }
+
+        // use 128 threads per block to maximimize gpu utilization
+        constexpr int threads_per_block = 128;
```
It looks like all these computationas are shared between forward/backward, so may be they can be abstracted away?

thorjohnsen replied:
Ok.

ngimel commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -627,19 +649,19 @@ Tensor host_softmax_backward(const Tensor &grad_, const Tensor &output_, int64_t
 
 Tensor log_softmax_cuda(const Tensor &input, const int64_t dim, const bool half_to_float){
-  return host_softmax<LogSoftMaxForwardEpilogue>(input, dim, half_to_float);
+  return host_softmax<LogSoftMaxForwardEpilogue,true>(input, dim, half_to_float);
```
I'm not too happy with 2 template arguments saying the same thing (I'm log! I'm not!), but since I did not come up with a not too ugly way to get rid of it, I'll let it slide.

ngimel reviewed (approved):
Some small changes, overall looks good.

ngimel commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -40,6 +40,8 @@ __global__ void softmax_warp_forward(output_t *dst, const input_t *src, int batc
             if (element_index < batch_element_count) {
                 elements[i][it] = src[i*element_count+it*WARP_SIZE];
+            } else {
+                elements[i][it] = -std::numeric_limits<acc_t>::infinity();
```
Why are you setting it to infinity the second time?

thorjohnsen committed:
Remove unnecessary init to infinity

thorjohnsen commented:
One of the softmax tests is still failing on accuracy, but it looks like that is due to a kluge in Pytorch. The test that fails runs softmax on a double precision tensor, but sees float like accuracy (i.e. relative error ~1e-8). The test was compiled with -D__HIP_PLATFORM_HCC__=1. My code uses WARP_SHFL_XOR for intrawarp reductions, if you look at the place where this is defined, there is a comment saying "HIP does not support double" and a specialization of WARP_SHFL_XOR that casts the value to a float before calling __shfl_xor. No test that uses WARP_SHFL_XOR will ever pass the double precision accuracy tests in test_nn.py because of this @iotamudelta.

https://github.com/pytorch/pytorch/blob/31e2d20c5eae46ea2bf1f24ddf96797adda860db/aten/src/THC/THCDeviceUtils.cuh#L63-L68

iotamudelta commented:
@thorjohnsen A double __shfl_xor(double, int, int) is defined for ROCm. This specialization should be removable.

thorjohnsen committed:
Remove double specialization of WARP_SHFL_XOR that fails accuracy tests

thorjohnsen commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,385 @@
+        }
+    }
+
+    constexpr uint32_t  FULL_MASK = 0xffffffff;
```
I will remove FULL_MASK.

thorjohnsen commented on aten/src/ATen/native/cuda/PersistentSoftmax.cuh:
```diff
@@ -0,0 +1,385 @@
+// Warp Softmax forward
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// WARP_BATCH number of batches.
```
I kind of inherited the comments. I agree they are not very clear, I'll try to improve that.

thorjohnsen commented:
71 tests passed, 4 failed. After Inspection, I don't think any of the 4 failed tests are related to this PR, but please don't take my word for it. Removing the double specialization of WARP_SHFL_XOR like you suggested fixed the failing Softmax test and a couple of other unrelated unit tests are also passing now, so we might have accidentally fixed a bug @iotamudelta. I will push the new code with your suggested improvements tomorrow morning @ngimel, then I think we should be ready for the merge.

thorjohnsen committed:
Address reviewer concerns

ngimel commented:
My comments are addressed, thank you.

thorjohnsen commented:
The 3 failed tests appear unrelated to this PR. One test failed with ImportError torch.onnx.symbolic_helper, the 2nd test failed with ImportError undefined symbol: _ZN2at19NonVariableTypeMode10is_enabledEv and the 3rd test failed because of a java.io.IOException: Backing channel 'JNLP4-connect connection from 209.249.227.2/209.249.227.2:43242' is disconnected. I believe this PR is ready to be merged.
